crates/ruff/src/rules/pyflakes/rules/repeated_keys.rs

```diff
@@ -125,88 +126,65 @@ impl Violation for MultiValueRepeatedKeyVariable {
     }
 }
 
-#[derive(Debug, Eq, PartialEq, Hash)]
-enum DictionaryKey<'a> {
-    Constant(ComparableConstant<'a>),
-    Variable(&'a str),
-}
-
-fn into_dictionary_key(expr: &Expr) -> Option<DictionaryKey> {
-    match expr {
-        Expr::Constant(ast::ExprConstant { value, .. }) => {
-            Some(DictionaryKey::Constant(value.into()))
-        }
-        Expr::Name(ast::ExprName { id, .. }) => Some(DictionaryKey::Variable(id)),
-        _ => None,
-    }
-}
-
 /// F601, F602
 pub(crate) fn repeated_keys(checker: &mut Checker, keys: &[Option<Expr>], values: &[Expr]) {
     // Generate a map from key to (index, value).
-    let mut seen: FxHashMap<DictionaryKey, FxHashSet<ComparableExpr>> =
+    let mut seen: FxHashMap<ComparableExpr, FxHashSet<ComparableExpr>> =
         FxHashMap::with_capacity_and_hasher(keys.len(), BuildHasherDefault::default());
 
     // Detect duplicate keys.
     for (i, key) in keys.iter().enumerate() {
         let Some(key) = key else {
             continue;
         };
-        if let Some(dict_key) = into_dictionary_key(key) {
-            if let Some(seen_values) = seen.get_mut(&dict_key) {
-                match dict_key {
-                    DictionaryKey::Constant(..) => {
-                        if checker.enabled(Rule::MultiValueRepeatedKeyLiteral) {
-                            let comparable_value: ComparableExpr = (&values[i]).into();
-                            let is_duplicate_value = seen_values.contains(&comparable_value);
-                            let mut diagnostic = Diagnostic::new(
-                                MultiValueRepeatedKeyLiteral {
-                                    name: checker.generator().expr(key),
-                                    repeated_value: is_duplicate_value,
-                                },
-                                key.range(),
-                            );
-                            if is_duplicate_value {
-                                if checker.patch(diagnostic.kind.rule()) {
-                                    diagnostic.set_fix(Fix::suggested(Edit::deletion(
-                                        values[i - 1].end(),
-                                        values[i].end(),
-                                    )));
-                                }
-                            } else {
-                                seen_values.insert(comparable_value);
+        let dict_key = ComparableExpr::from(key);
+        if let Some(seen_values) = seen.get_mut(&dict_key) {
+            let comparable_value: ComparableExpr = (&values[i]).into();
+            let is_duplicate_value = seen_values.contains(&comparable_value);
+            match dict_key {
+                Constant(_) | Tuple(_) | List(_) => {
```

Comment:
Not a complete list of everything that could be considered a literal, could add more to this perhaps. Depends on what we want to do with hashable vs unhashable types and if we want to keep the `_` default and say everything that isn't a literal must be a variable.

Reply:
Maybe this would be a good use case for `CheckableExprType` from https://github.com/astral-sh/ruff/pull/5602/files#diff-6dec723af30a5832377f43b03388b4c4fc4cd3897ef2e03e5511a17f7d3d1d09R146-R163

Reply:
I think enumerating here is ok -- only a small number of types are known to be hashable literals (e.g., you can't use a set here anyway), so I think it's fine to just enumerate constants and tuples.